### ISSUE TITLE
Bad performance with more than 100+ rows in an array field

### ISSUE THREAD

**Issue by pedro757** [TanStack/form#695]
### Describe the bug

Bad performance with 100+ rows in an array field

### Your minimal, reproducible example

https://codesandbox.io/p/sandbox/broken-silence-cwwxtp

### Steps to reproduce

1. Click `add 200` button
2. Click submit

### Expected behavior

I expect it to be faster, and not to see the `page unresponsive` alert of google chrome

### How often does this bug happen?

Every time

### Platform

*

### TanStack Form adapter

react-form

### TanStack Form version

0.19.4

**Comment by pedro757** [comment]
This happens even without validations and many other stuff

https://codesandbox.io/p/sandbox/xenodochial-cloud-87fvfz

**Comment by pedro757** [comment]
@crutchcorn  I'm currently using react-hook-form, I managed to make it work with 999+ without virtualization, but with a few caveats: 

1. Cannot access the `formState: { error }`, it slows down everything.
1. Cannot use any kind of validation of the library, No `<Controller />` with the `rules` prop and No `form.register(name, { validation: () => string})`
1. So I ended up just using `form.setValue()` and `form.getValues()` like this `<input value={form.getValues("name")} onChange={v => form.setValue("name", v)} />`
1. I do all validations when submitting the form manually, `form.handleSubmit(data => {
// HERE
})`

It feels slow with 999+, maybe virtualizing will improve that as [this comment suggests ](https://github.com/orgs/react-hook-form/discussions/7810#discussioncomment-2169754)

I can create a code sandbox this weekend.

**Comment by crutchcorn** [comment]
Well, to be fair that's not a very apt comparison, then. `Controller` is the headless comparison to `<form.Field>`. Moreover, `formState: {error}` is always the case with what we're validating.

**Comment by busy-mango** [comment]
I have the same problem.

**Comment by karur4n** [comment]
I had this problem too

**Comment by MinnDevelopment** [comment]
I ran into the same issue and ended up migrating to a map with an `order` array like this:

```js
{
  order: [UUID_1, UUID_2, ...],
  elements: {
   UUID_1: { ... },
   UUID_2: { ... },
   ...
  }
}
```

With this, I was able to avoid using an `index` that changes for almost every field when I insert a new value in the middle.

I can insert values by using the `setFieldValue` update function:

```ts
  function addCopy() {
    const copy = { ...item, id: uuidv4() };
    form.setFieldValue("items", (current) => ({
      order: current.order.toSpliced(index + 1, 0, copy.id),
      elements: {
        ...current.elements,
        [copy.id]: copy,
      },
    }));
  }
```

The item form fields are then simply ``<form.AppField name={`items.elements.${id}.text`}> ...``, which no longer has a changing `index`. This, in combination with memoized components, massively improved performance for my arrays.

**Comment by kaboofdotdev** [comment]
I'm also experiencing this issue when using the array field. I have hundreds of `form.AppField` elements on a page (think of a punch card), which unfortunately causes the browser tab to crash.

My current workaround is to avoid using `form.AppField` for those inputs and instead use a normal HTML input field mixed with the `getFieldValue` and `setFieldValue` methods from the form instance.

**Before**

```jsx
<form.AppField
    name="..."
    children={(field) => (
        <field.NumberInput />
    )}
/>
```

**After**

```jsx
<input
    type="number"
    value={form.getFieldValue("...")}
    onChange={(e) => form.setFieldValue("...", e.target.value)}
/>
```

Something about the `form.AppField` component is causing extreme performance issues.

**Comment by tzarger** [comment]
I am experiencing this issue too with form.AppField. using normal inputs for sure makes the form useable now. Is this an issue if we use the form.Field as well?

**Comment by stefan-huck** [comment]
> I am experiencing this issue too with form.AppField. using normal inputs for sure makes the form useable now. Is this an issue if we use the form.Field as well?

Yap, it is.

```
type Person = {
  surname: string
  name: string
  age: number
  email: string
}
```

Adding 80 persons with `field.pushValue`: 

```
<form.Field name={`people[${i}].surname`}>
  {subField => ( ... )}
```

causes `getFieldValue` to be called exponentially. See Video:

https://github.com/user-attachments/assets/f1374026-a949-468f-87f1-833db74b5ad8

In regard to @crutchcorn 's [reddit comment](https://www.reddit.com/r/reactjs/comments/1k29ta0/comment/mutvbeg) there might be some perf improvements in the future

**Comment by clayyclayyarend** [comment]
What's the likelihood this will be investigated? I just ran into this problem as well

**Comment by tzarger** [comment]
Is this going to be addressed any time soon? If not, will need to revert back to react-hook-form as I don't have any performance issues there with this form. Thanks! @crutchcorn

**Comment by mlivernoche** [comment]
I was having performance issues as well. My form has 3 text fields and 67 checkboxes, and a page can have multiples of these. Here's what I did to solve my issues:

- Installed the react compiler (easy for me)
- Used React Scan, this found that the culprit was shadcn's checkbox and label component. I changed those to HTML and this had a huge effect.

On my page, I place these forms in an accordion, and the initial render on my machine went from 600+ ms to 200ms. Renders after the initial (opening and closing the accordion) went from 300+ ms down to 16-32ms.

edit: also, these checkboxes and labels are inside an AppField.

**Comment by stefan-huck** [comment]
Since the latest version, the performance has been improved significantly!

Same code:
v1.27.4: https://stackblitz.com/edit/vitejs-vite-g4e9zmti?file=src%2FApp.tsx
v1.19.2: https://stackblitz.com/edit/vitejs-vite-7cfnfot4?file=src%2FApp.tsx

I know its not a real scenario:
On 1.27.4 adding 200 takes ~12 seconds, submitting that happens instant.
On 1.19.2 adding 200 takes ~40 seconds, submitting that takes more than a minute.

**Comment by stefan-huck** [comment]
While trying to migrate from RHF to TanStack-Form, I noticed a performance issue when calling `removeValue(0)` or `insertValue(0, { ... })` with ~100 active rows. 

Swapping, moving, clearing all items and appending a single item perform perfectly fine.
https://stackblitz.com/edit/vitejs-vite-n5p97woc?file=src%2FApp.tsx&terminal=dev

We are coming closer 🤞🍀

**Comment by Jack-Barry** [comment]
Ran into this today playing with an array of ~150 fields. Although the workaround [here](https://github.com/TanStack/form/issues/695#issuecomment-2978724362) is alright, would be nice to still be able to use the `AppField`/`Field` utilities without taking the performance hit though.

Seems to be worse calling `removeValue` on items closer to the beginning of the list than on items toward the end of it

**Comment by pedro757** [comment]
Hey there,

- Is it possible to use `@tanstack/react-virtual` for `<form.Field mode="array" ...>` ?
- Is there a working example of `@tanstack/react-form` and `@tanstack/react-virtual` working together?
- Would it improve the performance ?

**Comment by crutchcorn** [comment]
I'm happy to share that we've made some massive strides in performance as-of-late. In particular, this PR is likely to land sometime in the next 24 hours after we've triple-checked it against another maintainer's prod codebase:

https://github.com/TanStack/form/pull/2035

This PR, while not quite as fast as RHF yet, substantially moves the needle and no longer freezes the demo shared in OP.

This doesn't mean our work is done yet, however. We have one more massive PR internals rework that's nebulously passing RHF in some benchmarks we have (but still loses in many real-world usecases, so mixed bag).

We're looking into how far we can push perf before a future v2 release when some breaking changes in userland will enable much more perf work.

We appreciate you all being patient with us as we grow in this avenue!

**Comment by tzarger** [comment]
@crutchcorn you mentioned that PR would land in the next 24 hours.. but it has been 2 weeks ago... and the PR is still Open ... Do you have any new dates for the PR to land and released?

Thank you for the effort!

**Comment by crutchcorn** [comment]
Yes, sorry @tzarger. We wanted to release it very quickly, but wanted to test it more thoroughly and communicate with other TanStack libraries about some of the details about the internals changes.

... And then I got swamped at work 🤡 Sorry all! Out now!

I'm going to close this issue as the original repro at least doesn't crash the browser. We're far from done with performance improvements, but I think per the scope of this issue we've made massive strides.

**Comment by timker** [comment]
I can say 1.23.24 has improved loading perf by way over %50. Amazing!

removing rows, is way faster but still seems slower than expected.

I'm using lit-form with repeat directive, e.g.

  ${repeat(rowField.state.value, 
                    (_, index) => index, 
                    (_, index) => this.renderRow(formController, index))
                  }
